### PR TITLE
[WIP][Model] Add framework-native MOSS-TTS-Realtime support

### PR DESCRIPTION
**Pull request by CloudRipple**
## Motivation

`OpenMOSS-Team/MOSS-TTS-Realtime` is designed for interactive voice agents:
text or token deltas can arrive while a turn is active, audio can be emitted
before the producer finishes sending text, and successful turns can retain
text/audio KV context for later turns in the same session.

The existing speech APIs can stream audio for an already complete input, but
they cannot append new LLM SSE deltas to one live model request, park an
input-starved turn without discarding its state, or continue a later turn from
the same warm session.

This PR adds a model-owned bidirectional realtime path while keeping the
ordinary complete-input `/v1/audio/speech` API available for offline synthesis,
evaluation, and OpenAI-compatible clients.

## Modifications

- Add `WS /v1/audio/speech/realtime` with typed session and turn lifecycle
  events, ordered `input.text` and `input.tokens` updates, acknowledgements,
  `input.done`, cancellation, streamed 24 kHz PCM, and explicit session close.
- Add a dedicated realtime coordinator and input-update control plane, including
  registration barriers, configured input-stage routing, idempotent updates,
  terminal tombstones, and cancellation-safe cleanup.
- Extend scheduling for active incremental input: row materialization,
  ready/parked partitioning, wakeup, mixed-rate batching, bounded pending input,
  timeout handling, and one-time terminal finalization.
- Add transactional warm-session support. Successful turns commit canonical
  mixed text/audio history and retain KV state; cache loss triggers deterministic
  replay, while cancellation, failure, explicit close, TTL expiry, and shutdown
  release host and physical ownership.
- Integrate the checkpoint with SGLang's Qwen3 backbone, paged KV cache,
  Radix/Chunk cache infrastructure, mixed text/audio embeddings, exact
  request-owned sampling, and an HF-compatible local RVQ transformer.
- Add complete-input speech support with ephemeral offline sessions, preserved
  model-owned sampling defaults, context-derived generation limits, and no
  arbitrary turn-frame cap.
- Add a persistent 16-codebook streaming vocoder with fixed slots, BF16 native
  SDPA by default, FP32 fallback, exact reset/reuse semantics, and dense CUDA
  Graph capture for queued frame counts 1 through 12.
- Add bounded reference-audio caching, content-aware local-file revalidation,
  same-key single-flight, cross-request codec microbatching, and per-item retry.
- Add resource limits and observability for active turns, logical and physical
  sessions, held KV, model-state rows, codec slots, cleanup outcomes, and the
  first-PCM critical path.
- Add a colocated example config, a usage cookbook, and model/coordinator/
  scheduler/WebSocket/session/codec/concurrency regression coverage.

Current behavioral boundaries:

- generation starts after the checkpoint exposes 12 stable BPE tokens, or when
  `input.done` closes a shorter input;
- `input.done` is terminal for one turn, but a later turn can reuse the same
  committed session;
- one turn can be active per session, with 16 active turns and codec slots by
  default across sessions;
- realtime output is signed 16-bit mono PCM at 24 kHz.

## Related Issues

- Upstream model card:
  https://github.com/OpenMOSS/MOSS-TTS/blob/main/docs/moss_tts_realtime_model_card.md

## Checklist

- [ ] Format your code according with pre-commit.
- [ ] Add unit tests.
- [ ] Update documentation / docstrings / example tutorials as needed.
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the
`run-ci` label. Once labeled, every subsequent push re-triggers CI as
long as the label remains. Use `/tag-and-rerun-ci higgs` or
`/tag-and-rerun-ci moss` to select a TTS CI model, and
`/tag-and-rerun-ci fun-asr` or `/tag-and-rerun-ci qwen3-asr` to select an ASR
CI model. One selector from each family can be combined, for example
`/tag-and-rerun-ci moss fun-asr`. Draft PRs are skipped even if labeled.

### CODE REVIEW

**Review by luojiaxuan — COMMENTED** [review]
Issue 1812 has a fixed-code A/B reproducer; the codec slot lifetime needs to follow session_id rather than request_id.

File: sglang_omni/models/moss_tts_realtime/streaming_vocoder.py [hunk 1]
```diff
@@ -0,0 +1,1206 @@
...
+        if self._session is not None:
+            self._session.close()
+            self._session = None
+            self._resource_totals["codec_session_close_total"] += 1
+            self._emit_codec_event("codec-session", "session_close")
+
+    def create_stream_state(self, request_id: str) -> _RealtimeStreamState:
+        try:
+            session = self._ensure_session_graphed()
+            slot = session.acquire()
```
Comment (luojiaxuan) [inline sglang_omni/models/moss_tts_realtime/streaming_vocoder.py:858]: Please key this lease by session_id and reuse it for later turns in the same session. A new request_id or turn_id should not start fresh codec state.

File: sglang_omni/models/moss_tts_realtime/streaming_vocoder.py [hunk 2]
```diff
@@ -0,0 +1,1206 @@
...
+        if state.pending:
+            if state.slot is None:
+                raise RuntimeError("MOSS-TTS-Realtime final flush has no codec slot")
+            session = self._ensure_session()
+            while state.pending:
+                step_frames = min(len(state.pending), _STEADY_CHUNK_FRAMES)
+                codes = torch.stack(state.pending[:step_frames], dim=1)
+                del state.pending[:step_frames]
+                audio_parts.append(session.step({state.slot: codes})[state.slot])
+        self._release_state_slot(request_id, state)
```
Comment (luojiaxuan) [inline sglang_omni/models/moss_tts_realtime/streaming_vocoder.py]: Do not release the slot after a successful turn final flush. Drain pending PCM here, then release and reset only on session close, TTL expiry, abort, or failure.

**Comment by luojiaxuan** [comment]
Fixed-code A/B for the same 60 second sample range is in issue 1812: https://github.com/sgl-project/sglang-omni/issues/1812

Before, per-turn codec reset: https://github.com/luojiaxuan/S2S_omni/releases/download/talk110-codec-context-ab-20260829/talk110_boundary109_60s_reset.wav

Expected after fix, codec context preserved across turns: https://github.com/luojiaxuan/S2S_omni/releases/download/talk110-codec-context-ab-20260829/talk110_boundary109_60s_continuous.wav

File: sglang_omni/models/moss_tts_realtime/streaming_vocoder.py (hunk 1 above)
Comment (CloudRipple) [reply sglang_omni/models/moss_tts_realtime/streaming_vocoder.py:858]: Done in `9f104f9e` (now on the rebased branch, HEAD `acfd4374`). Codec slots are now leased per `session_id` via `_codec_sessions`; chunk metadata carries `session_id`/`turn_id`, so a new `request_id`/`turn_id` in the same session reuses the held slot instead of starting fresh codec state. A successful turn only drains pending PCM.

Live evidence on H200: a six-turn 78s session shows codec `acquire_total=1` across all six turns (single shared slot); `release_total=1` lands only after `session.close`, with `held=0` / `free=80` afterwards — no leak. An ephemeral `stream=true` request still does acquire/release +1/+1 immediately.

File: sglang_omni/models/moss_tts_realtime/streaming_vocoder.py (hunk 2 above)
Comment (CloudRipple) [reply sglang_omni/models/moss_tts_realtime/streaming_vocoder.py]: Done. The final-flush path now only drains pending PCM; release+reset happens exactly at the four teardown points listed here:

1. session close — explicit admin close targeting both engine and vocoder, plus a stream-edge marker for ephemeral requests; a close arriving mid-drain marks the session `closing` and the release lands at that request's teardown;
2. idle TTL sweep — 2x the engine-side TTL, piggybacked on the regular sweep;
3. abort, including `turn.cancel`;
4. decode failure.

Default slot capacity is raised 16 → 80 (`max_held_sessions=64` + `max_active_turns=16`), and the colocated GPU-memory budget derives from the same limits. `resource_snapshot` now exposes `codec_held_sessions` / `codec_closing_sessions` / `codec_session_idle_ttl_s`, and codec events carry `session_id`.
